daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala

```diff
@@ -100,4 +104,112 @@ object Util {
       case None => Right(None)
     }
 
+  /** Compute the inputs to a DAML transaction (in traversal order), that is, the referenced contracts, parties,
+    *  and key.
+    */
+  def collectInputs(
+      tx: Transaction.Transaction,
+      addContract: ContractId => Unit,
+      addParty: Ref.Party => Unit,
+      addGlobalKey: GlobalKey => Unit,
+  ): Unit = {
+
+    val localContract = tx.localContracts
+
+    def addContractInput(coid: ContractId): Unit =
+      if (!localContract.isDefinedAt(coid))
+        addContract(coid)
+
+    def addPartyInputs(parties: Set[Ref.Party]): Unit =
+      parties.toList.sorted(Ref.Party.ordering).foreach(addParty)
+
+    def addContractKey(
+        tmplId: Ref.Identifier,
+        key: Node.KeyWithMaintainers[Value[ContractId]],
+    ): Unit =
+      addGlobalKey(GlobalKey.assertBuild(tmplId, key.key))
+
+    tx.foreach { case (_, node) =>
+      node match {
+        case _: Node.NodeRollback[_] =>
+          // TODO https://github.com/digital-asset/daml/issues/8020
+          sys.error("rollback nodes are not supported")
+        case fetch: Node.NodeFetch[Value.ContractId] =>
+          addContractInput(fetch.coid)
+          fetch.key.foreach(addContractKey(fetch.templateId, _))
+
+        case create: Node.NodeCreate[Value.ContractId] =>
+          create.key.foreach(addContractKey(create.templateId, _))
+
+        case exe: Node.NodeExercises[NodeId, Value.ContractId] =>
+          addContractInput(exe.targetCoid)
+          exe.key.foreach(addContractKey(exe.templateId, _))
```

Comment:
Why do you have to add the key to the inputs if the `byKey` flag is not set?

Reply:
We don't. If the `byKey` flag is not set, then `exe.key = None`.

Reply:
Let's postpone semantics consideration to another PR. 
For now, I just move the code around keeping as much as possible the original behavior of the utility as defined in KV.